srk/src/transitionSystem.ml

```diff
@@ -662,18 +684,158 @@ module Make
     in
     List.map invariants (L.all_loops (L.loop_nest tg))
 
-  let simplify p tg =
+
+  let inline ?(depth=(-1)) tg entry (displayer: t -> unit) (assertion_map : 'x SrkUtil.Int.Map.t) = 
+    let pmap = PHT.create 998 in (* y in pmap[x] means call from x->y *)
+    let qmap = PHT.create 998 in (* (x, z) in qmap[y] means call from x->y via call-edge z *)
+    let greatest = ref (WG.fold_vertex max tg (-1)) in
+    let assertions = ref assertion_map in   
+    PHT.add pmap (entry, -1) PS.empty;
+    PHT.add qmap (entry, -1) PPS.empty;
+    let procedures = 
+      WG.fold_edges (fun (_, w, _) acc -> 
+        match w with 
+        | Call (x, y) -> 
+          begin match PHT.find_opt pmap (x, y) with 
+          | None -> 
+            PHT.add pmap (x, y) PS.empty 
+          | Some _ -> ()
+          end;
+          begin match PHT.find_opt qmap (x, y) with 
+          | None -> 
+            PHT.add qmap (x, y) PPS.empty 
+          | Some _ -> () 
+          end;
+          PS.add (x, y) acc
+        | _ -> acc 
+        ) tg (PS.add (entry, -1) PS.empty) in 
+    let rec dfs (proc: int * int) tg src (f: (vertex * vertex) -> vertex -> vertex * 'a label * vertex -> unit) (visited : ISet.t) = 
+      WG.fold_succ_e (fun (u, w, v) visited -> 
+        f proc src (u, w, v);
+        begin match ISet.find_opt v visited with 
+            | None -> ISet.union (dfs proc tg v f (ISet.add v visited)) visited 
+            | Some _ -> visited 
+            end) tg src visited in 
+      PS.iter (fun (x, y) -> 
+        ignore @@ dfs (x, y) tg x (fun proc src (_, w, v) -> 
+          match w with 
+          | Call (x, y) -> 
+            let caller_set = PHT.find pmap proc in 
+            let callee_set = PHT.find qmap (x, y) in
+            PHT.add pmap proc (PS.add (x, y) caller_set);
+            PHT.add qmap (x, y) (PPS.add (proc, (src, v)) callee_set);
+          | _ -> () 
+          ) (ISet.add x ISet.empty)) procedures; (* populate pmap, qmap *)
+      let compute_sinks () = 
```

Comment:
(Reverse) topological traversal of the callgraph?  OCamlgraph has a Topological functor that accepts WG.RecGraph.Callgraph as a parameter.

Reply:
Or better yet, use the Loop module in srk, which will also identify connected components in the callgraph which you shouldn't inline.